libs/vkd3d/command.c

```diff
@@ -8345,121 +8345,559 @@ static void STDMETHODCALLTYPE d3d12_command_list_CopyTiles(d3d12_command_list_if
+    if (dst_resource->format->vk_aspect_mask & (VK_IMAGE_ASPECT_DEPTH_BIT | VK_IMAGE_ASPECT_STENCIL_BIT))
+        return dst_resource->format;
+
+    if (dst_resource->format->type != VKD3D_FORMAT_TYPE_TYPELESS)
+        return dst_resource->format;
+
+    if (src_resource->format->type != VKD3D_FORMAT_TYPE_TYPELESS)
+        return src_resource->format;
+
+    return vkd3d_format_from_d3d12_resource_desc(list->device, &dst_resource->desc, format);
+}
+
+enum vkd3d_resolve_path d3d12_command_list_select_resolve_path(struct d3d12_command_list *list,
+        struct d3d12_resource *dst_resource, struct d3d12_resource *src_resource, uint32_t region_count,
+        const VkImageResolve2 *regions, DXGI_FORMAT format, D3D12_RESOLVE_MODE mode)
+{
+    const struct vkd3d_format *vkd3d_format = d3d12_command_list_get_resolve_format(list, dst_resource, src_resource, format);
+    enum vkd3d_resolve_path path;
     unsigned int i;
 
-    if (mode != D3D12_RESOLVE_MODE_AVERAGE)
+    if (dst_resource->format->vk_aspect_mask != src_resource->format->vk_aspect_mask)
     {
-        FIXME("Resolve mode %u is not yet supported.\n", mode);
-        return;
+        /* Mismatched aspects may happen with some typeless formats */
+        path = VKD3D_RESOLVE_PATH_RENDER_PASS_PIPELINE;
+    }
+    else if (vkd3d_format->vk_aspect_mask & (VK_IMAGE_ASPECT_DEPTH_BIT | VK_IMAGE_ASPECT_STENCIL_BIT))
+    {
+        /* The direct path only supports color images with the AVERAGE mode */
+        path = VKD3D_RESOLVE_PATH_RENDER_PASS_ATTACHMENT;
     }
+    else if (mode != D3D12_RESOLVE_MODE_AVERAGE)
+    {
+        /* Vulkan only supports SAMPLE_ZERO for color images for which D3D12 does
+         * not even have an equivalent, we can only implement this in shaders. */
+        path = VKD3D_RESOLVE_PATH_RENDER_PASS_PIPELINE;
+    }
+    else if (dst_resource->format->type == VKD3D_FORMAT_TYPE_TYPELESS ||
+            src_resource->format->type == VKD3D_FORMAT_TYPE_TYPELESS)
+    {
+        /* Prefer the direct path even for typeless formats if possible */
+        if (vkd3d_format->vk_format == dst_resource->format->vk_format &&
+                vkd3d_format->vk_format == src_resource->format->vk_format)
+            path = VKD3D_RESOLVE_PATH_DIRECT;
+        else
+            path = VKD3D_RESOLVE_PATH_RENDER_PASS_ATTACHMENT;
+    }
+    else
+        path = VKD3D_RESOLVE_PATH_DIRECT;
 
-    if (mode == D3D12_RESOLVE_MODE_AVERAGE && (dst_resource->format->vk_aspect_mask & VK_IMAGE_ASPECT_DEPTH_BIT))
+    /* The attachment path has a number of restrictions that may require us to fall
+     * back to the shader-based path. */
+    if (path == VKD3D_RESOLVE_PATH_RENDER_PASS_ATTACHMENT)
     {
-        FIXME("AVERAGE resolve on DEPTH aspect is not supported yet.\n");
-        return;
+        if (!(src_resource->desc.Flags & (D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET | D3D12_RESOURCE_FLAG_ALLOW_DEPTH_STENCIL)))
```

Comment:
Shouldn't this apply to dst_resource as well?

Reply:
We check for that later; source image is fine only with `SAMPLED_IMAGE` usage, which all multisampled images have.